### PR TITLE
`mvn spotless:apply` now prints out statistics

### PR DESCRIPTION
**Pull request by blacelle**
Follows https://github.com/diffplug/spotless/pull/1506

### CODE REVIEW

**Comment by blacelle** [comment]
Here is another step forward. I introduce a Formatter.name property that I'm not 100% comfortable with.

The log now looks like:

```
[INFO] ------------------------------------------------------------------------
[INFO] Building Spotless Maven Plugin Tests 1.0.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- spotless-maven-plugin:2.31.0-SNAPSHOT:apply (default-cli) @ spotless-maven-plugin-tests ---
[INFO] Writing clean file: /private/var/folders/8b/p64c8tfs4d7gf3v8tcmwbz580000gn/T/junit17574313710916441137/json_test.json
[INFO] Spotless.Json is keeping 2 files clean - 1 were changed to be clean, 1 were already clean, 0 were skipped because caching determined they were already clean
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
```

I was keen to add the equivalent log in Gradle, but I'm too weak in Gradle. If this is a relevant improvment, I would appreciate a pointer to do so.

File: lib/src/main/java/com/diffplug/spotless/Formatter.java [hunk 1]
```diff
@@ -103,6 +112,8 @@ public static Formatter.Builder builder() {
 	}
 
 	public static class Builder {
+		// optional parameters
+		private String name = "misc";
```
Comment (nedtwigg) [inline lib/src/main/java/com/diffplug/spotless/Formatter.java:116]: I would call this "unnamed formatter" rather "misc".

File: plugin-maven/src/main/java/com/diffplug/spotless/maven/ImpactedFilesTracker.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Tracks the number of processed files, typically by a single Formatter for a whole repository
+ */
+public class ImpactedFilesTracker {
+	protected final AtomicInteger nbSkipped = new AtomicInteger();
+	protected final AtomicInteger nbChecked = new AtomicInteger();
+	protected final AtomicInteger nbCleaned = new AtomicInteger();
```
Comment (nedtwigg) [inline plugin-maven/src/main/java/com/diffplug/spotless/maven/ImpactedFilesTracker.java:26]: A single `FormatterStep` does not support multithreading, so our `Formatter` doesn't either. I think regular integers with `++` is better to make it clear that the code is not actually threadsafe.

File: plugin-maven/src/main/java/com/diffplug/spotless/maven/SpotlessApplyMojo.java [hunk 3]
```diff
@@ -33,25 +33,39 @@ public class SpotlessApplyMojo extends AbstractSpotlessMojo {
...
 			}
 
 			upToDateChecker.setUpToDate(file.toPath());
 		}
+
+		// We print the number of considered files which is useful when ratchetFrom is setup
+		int nbSkipped = impactedFilesTracker.getSkipped();
+		int nbChecked = impactedFilesTracker.getChecked();
+		int nbCleaned = impactedFilesTracker.getCleaned();
+		int totalProcessed = nbSkipped + nbChecked + nbCleaned;
```
Comment (nedtwigg) [inline plugin-maven/src/main/java/com/diffplug/spotless/maven/SpotlessApplyMojo.java:67]:
```
int totalProcessed = nbSkipped + nbChecked
int numAlreadyClean = nbChecked - nbCleaned
```

An alternative which might be clearer is to change the fields in `ImpactedFilesTracker`, e.g.

```java
if (upToDateChecker.isUpToDate(file.toPath())) {
  impactedFilesTracker.skippedAsCleanCache();
}
PaddedCell.DirtyState dirtyState = PaddedCell.calculateDirtyState(formatter, file);
if (!dirtyState.isClean() && !dirtyState.didNotConverge()) {
  impactedFilesTracker.cleaned()
} else {
  impactedFilesTracker.checkedButAlreadyClean()
}
```

With this formulation, `totalProcessed = skippedAsCleanCache + cleaned + checkedButAlreadyClean`

**Review by nedtwigg — CHANGES_REQUESTED** [review]
Thanks for the change, it's close to mergeable.

Re: the Gradle plugin, it's tough because the up-to-date checking happens within the gradle machinery, and I'm not aware of a way to get a count of how many were skipped due to up-to-date checking. Regardless, they're not coupled, so it can always be added later.

File: lib/src/main/java/com/diffplug/spotless/Formatter.java (hunk 1 above)
Comment (blacelle) [reply lib/src/main/java/com/diffplug/spotless/Formatter.java:116]: I suggest _unnamed_ as _unnamed formatter_ looks a bit redundant.

File: plugin-maven/src/main/java/com/diffplug/spotless/maven/SpotlessApplyMojo.java (hunk 3 above)
Comment (blacelle) [reply plugin-maven/src/main/java/com/diffplug/spotless/maven/SpotlessApplyMojo.java]: I went for the second suggestion. It was the initial spirit (while the execution was... lacking rigour).

File: lib/src/main/java/com/diffplug/spotless/Formatter.java (hunk 1 above)
Comment (nedtwigg) [reply lib/src/main/java/com/diffplug/spotless/Formatter.java]: Sounds good.

**Review by nedtwigg — APPROVED** [review]
LGTM. Add an entry under `###Added` in `plugin-maven/CHANGELOG` and I'll press merge.

**Comment by blacelle** [comment]
README entry added. I also added a condition so that synthesis kicks-in only if `totalProcessed>0`

**Comment by nedtwigg** [comment]
I changed the `debug` to `warn`, as it's likely a user mistake to have a formatter which is doing nothing. Example such mistake

- #437

Are these changes okay with you? If so I will merge.

**Comment by blacelle** [comment]
Thanks for the improvments.

**Comment by blacelle** [comment]
@nedtwigg I'm not sure about the WARN if there is not a single matching file. As the maven-plugin is executed per module, we may typically have a `<json>` section in the parent multi-module, and have some sub-module not having a single json to format.

**Comment by nedtwigg** [comment]
A fair amount of traffic on our bug tracker is

- "it's not formatting any files"
- "did you check the target includes"
- "oh"

I'm happy to backtrack on this now or in the future, but it seems like it would be better to have a way for child pom to say e.g. `<spotless><someFormat><disable>true` to turn off warnings in that subproject.

**Comment by blacelle** [comment]
@nedtwigg I'm also a user encountering many "it's not formatting any files". Which is one motivation for this PR. However, I foresee many WARNs in my own builds. I'd like to run a SNAPSHOT with one of my repository (to get a real feedback): do you know how I can deploy a SNAPSHOT in mvn local repository with the new setup ? (based on https://www.benediktritter.de/maven-plugin-development/)

**Comment by nedtwigg** [comment]
I think `publishToMavenLocal` should still work with the new `maven-plugin-development` build

- https://github.com/diffplug/spotless/blob/main/CONTRIBUTING.md#maven

But that's a good point, I think we can use JITPACK again now too. I just turned JITPACK building back on, and it looks like it's working okay as of current master `4c7b616e510bd83167ff329361ad819fac28ae6e`

https://jitpack.io/com/github/diffplug/spotless/4c7b616e510bd83167ff329361ad819fac28ae6e/build.log

**Comment by nedtwigg** [comment]
Ah, JITPACK is broken, working on that...

**Comment by blacelle** [comment]
> ./gradlew publishToMavenLocal

fails with:

```
spotless ./gradlew publishToMavenLocal

> Configure project :plugin-gradle
Signing plugin detected. Will automatically sign the published artifacts.

> Task :plugin-gradle:signPluginMavenPublication FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':plugin-gradle:signPluginMavenPublication'.
> Cannot perform signing task ':plugin-gradle:signPluginMavenPublication' because it has no configured signatory

* Try:
> Run with --stacktrace option to get the stack trace.
> Run with --info or --debug option to get more log output.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.

You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.

See https://docs.gradle.org/7.6/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 1s
41 actionable tasks: 9 executed, 32 up-to-date

Publishing build scan...
https://gradle.com/s/h4jvq7rusebeg
```

Same for `./gradlew maven-plugin:publishToMavenLocal`

```
cd maven-plugin
../gradlew publishToMavenLocal
```

is OK.

---

A run of ` mvn com.diffplug.spotless:spotless-maven-plugin:2.32.0-SNAPSHOT:apply` over https://github.com/solven-eu/pepper gives:

```
[INFO] --------------< io.github.solven-eu.pepper:pepper-parent >--------------
[INFO] Building pepper-parent 4.3-SNAPSHOT                               [1/25]
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- spotless-maven-plugin:2.32.0-SNAPSHOT:apply (default-cli) @ pepper-parent ---
[WARNING] Spotless.Format has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[WARNING] Spotless.Markdown has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[WARNING] Spotless.Json has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[WARNING] Spotless.Java has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] Sorting file /var/folders/8b/p64c8tfs4d7gf3v8tcmwbz580000gn/T/pom10821452448486918927.xml
[INFO] Pom file is already sorted, exiting
[INFO] Spotless.Pom is keeping 1 files clean - 0 were changed to be clean, 1 were already clean, 0 were skipped because caching determined they were already clean
[INFO] 
[INFO] --------------< io.github.solven-eu.pepper:pepper-static >--------------
[INFO] Building pepper-static 4.3-SNAPSHOT                               [2/25]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- spotless-maven-plugin:2.32.0-SNAPSHOT:apply (default-cli) @ pepper-static ---
[WARNING] Spotless.Markdown has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[WARNING] Spotless.Format has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] Sorting file /var/folders/8b/p64c8tfs4d7gf3v8tcmwbz580000gn/T/pom9606408092992840722.xml
[INFO] Pom file is already sorted, exiting
[INFO] Spotless.Pom is keeping 1 files clean - 0 were changed to be clean, 1 were already clean, 0 were skipped because caching determined they were already clean
[WARNING] Spotless.Java has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[WARNING] Spotless.Json has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] 
[INFO] ------------< io.github.solven-eu.pepper:aggregator-pepper >------------
[INFO] Building aggregator-pepper 4.3-SNAPSHOT                           [3/25]
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- spotless-maven-plugin:2.32.0-SNAPSHOT:apply (default-cli) @ aggregator-pepper ---
[INFO] Sorting file /var/folders/8b/p64c8tfs4d7gf3v8tcmwbz580000gn/T/pom11573418929783516858.xml
[INFO] Pom file is already sorted, exiting
[INFO] Spotless.Pom is keeping 1 files clean - 0 were changed to be clean, 1 were already clean, 0 were skipped because caching determined they were already clean
[WARNING] Spotless.Java has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] Spotless.Json is keeping 5 files clean - 0 were changed to be clean, 5 were already clean, 0 were skipped because caching determined they were already clean
[INFO] Spotless.Markdown is keeping 3 files clean - 0 were changed to be clean, 3 were already clean, 0 were skipped because caching determined they were already clean
[WARNING] Spotless.Format has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] 
[INFO] --------------< io.github.solven-eu.pepper:pepper-agent >---------------
[INFO] Building pepper-agent 4.3-SNAPSHOT                                [4/25]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- spotless-maven-plugin:2.32.0-SNAPSHOT:apply (default-cli) @ pepper-agent ---
[INFO] Spotless.Java is keeping 8 files clean - 0 were changed to be clean, 8 were already clean, 0 were skipped because caching determined they were already clean
[WARNING] Spotless.Json has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] Sorting file /var/folders/8b/p64c8tfs4d7gf3v8tcmwbz580000gn/T/pom16472238603268760438.xml
[INFO] Pom file is already sorted, exiting
[INFO] Spotless.Pom is keeping 1 files clean - 0 were changed to be clean, 1 were already clean, 0 were skipped because caching determined they were already clean
[WARNING] Spotless.Format has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] Spotless.Markdown is keeping 1 files clean - 0 were changed to be clean, 1 were already clean, 0 were skipped because caching determined they were already clean
...
```

This is many warnings in my perspective.

Regarding people with wrong includes: what's often lacking to me (with spotless) is proper loggings (e.g. this PR adds a DEBUG log when cleaning a file). I'm not sure (...) people with issues would try `mvn ... -X`, but as of now, even those trying debug log does not get much to digest.

**Comment by blacelle** [comment]
A run of ` mvn com.diffplug.spotless:spotless-maven-plugin:2.32.0-SNAPSHOT:apply` over https://github.com/solven-eu/pepper gives:

```
[INFO] --------------< io.github.solven-eu.pepper:pepper-parent >--------------
[INFO] Building pepper-parent 4.3-SNAPSHOT                               [1/25]
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- spotless-maven-plugin:2.32.0-SNAPSHOT:apply (default-cli) @ pepper-parent ---
[WARNING] Spotless.Format has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[WARNING] Spotless.Markdown has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[WARNING] Spotless.Json has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[WARNING] Spotless.Java has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] Sorting file /var/folders/8b/p64c8tfs4d7gf3v8tcmwbz580000gn/T/pom10821452448486918927.xml
[INFO] Pom file is already sorted, exiting
[INFO] Spotless.Pom is keeping 1 files clean - 0 were changed to be clean, 1 were already clean, 0 were skipped because caching determined they were already clean
[INFO] 
[INFO] --------------< io.github.solven-eu.pepper:pepper-static >--------------
[INFO] Building pepper-static 4.3-SNAPSHOT                               [2/25]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- spotless-maven-plugin:2.32.0-SNAPSHOT:apply (default-cli) @ pepper-static ---
[WARNING] Spotless.Markdown has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[WARNING] Spotless.Format has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] Sorting file /var/folders/8b/p64c8tfs4d7gf3v8tcmwbz580000gn/T/pom9606408092992840722.xml
[INFO] Pom file is already sorted, exiting
[INFO] Spotless.Pom is keeping 1 files clean - 0 were changed to be clean, 1 were already clean, 0 were skipped because caching determined they were already clean
[WARNING] Spotless.Java has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[WARNING] Spotless.Json has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] 
[INFO] ------------< io.github.solven-eu.pepper:aggregator-pepper >------------
[INFO] Building aggregator-pepper 4.3-SNAPSHOT                           [3/25]
[INFO] --------------------------------[ pom ]---------------------------------
[INFO] 
[INFO] --- spotless-maven-plugin:2.32.0-SNAPSHOT:apply (default-cli) @ aggregator-pepper ---
[INFO] Sorting file /var/folders/8b/p64c8tfs4d7gf3v8tcmwbz580000gn/T/pom11573418929783516858.xml
[INFO] Pom file is already sorted, exiting
[INFO] Spotless.Pom is keeping 1 files clean - 0 were changed to be clean, 1 were already clean, 0 were skipped because caching determined they were already clean
[WARNING] Spotless.Java has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] Spotless.Json is keeping 5 files clean - 0 were changed to be clean, 5 were already clean, 0 were skipped because caching determined they were already clean
[INFO] Spotless.Markdown is keeping 3 files clean - 0 were changed to be clean, 3 were already clean, 0 were skipped because caching determined they were already clean
[WARNING] Spotless.Format has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] 
[INFO] --------------< io.github.solven-eu.pepper:pepper-agent >---------------
[INFO] Building pepper-agent 4.3-SNAPSHOT                                [4/25]
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- spotless-maven-plugin:2.32.0-SNAPSHOT:apply (default-cli) @ pepper-agent ---
[INFO] Spotless.Java is keeping 8 files clean - 0 were changed to be clean, 8 were already clean, 0 were skipped because caching determined they were already clean
[WARNING] Spotless.Json has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] Sorting file /var/folders/8b/p64c8tfs4d7gf3v8tcmwbz580000gn/T/pom16472238603268760438.xml
[INFO] Pom file is already sorted, exiting
[INFO] Spotless.Pom is keeping 1 files clean - 0 were changed to be clean, 1 were already clean, 0 were skipped because caching determined they were already clean
[WARNING] Spotless.Format has no target files. Examine your `<includes>`: https://github.com/diffplug/spotless/tree/main/plugin-maven#quickstart
[INFO] Spotless.Markdown is keeping 1 files clean - 0 were changed to be clean, 1 were already clean, 0 were skipped because caching determined they were already clean
...
```

This is many warnings in my perspective.

Regarding people with wrong includes: what's often lacking to me (with spotless) is proper loggings (e.g. this PR adds a DEBUG log when cleaning a file). I'm not sure (...) people with issues would try `mvn ... -X`, but as of now, even those digging into debug log does not get much to digest.

**Comment by nedtwigg** [comment]
I am not familiar with Maven culture. Defining a bunch of goals in the parent pom to apply to children but don't have any effect seems like bad practice. But I'm not a maven user, and I'm happy to merge whatever native maven users feel the default behavior ought to be.

**Comment by blacelle** [comment]
> Defining a bunch of goals in the parent pom to apply to children but don't have any effect seems like bad practice.

In my perspective, this is totally legit and standard practise. But I'm just a single user. Do you know any other mvn Spotless user which may give his opinion on this?

**Comment by nedtwigg** [comment]
I always lean towards PR author's preference, merged :)

**Comment by nedtwigg** [comment]
Published in `2.32.0`.

**Pull request merged**